presto-native-execution/presto_cpp/presto_protocol/WriteProtocol.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/presto_protocol/WriteProtocol.h"
+
+#include "velox/connectors/hive/HiveConnector.h"
+#include "velox/connectors/hive/HiveWriteProtocol.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::connector;
+using namespace facebook::velox::connector::hive;
+using namespace facebook::velox::exec;
+
+namespace facebook::presto::protocol {
+
+namespace {
+
+std::string makePartitionUpdate(
+    const std::shared_ptr<const WriterParameters> writerParameters,
+    vector_size_t numWrittenRows) {
+  const auto hiveWriterParameters =
+      std::dynamic_pointer_cast<const HiveWriterParameters>(writerParameters);
+  VELOX_CHECK(
+      hiveWriterParameters != nullptr,
+      "Presto commit protocol only supports Hive commits at the moment.")
+  // clang-format off
+  auto partitionUpdateJson = folly::toJson(
+   folly::dynamic::object
+     ("name", "")(
+      "updateMode",
+      HiveWriterParameters::updateModeToString(
+          hiveWriterParameters->updateMode()))("writePath", hiveWriterParameters->writeDirectory())
+     ("targetPath", hiveWriterParameters->targetDirectory())
+     ("fileWriteInfos", folly::dynamic::array(
+       folly::dynamic::object
+         ("writeFileName", hiveWriterParameters->writeFileName())
+         ("targetFileName", hiveWriterParameters->targetFileName())
+         ("fileSize", 0)))
+     ("rowCount", numWrittenRows)
+     ("inMemoryDataSizeInBytes", 0)
+     ("onDiskDataSizeInBytes", 0)
+     ("containsNumberedFileNames", true));
+  // clang-format on
+  return partitionUpdateJson;
+}
+
+RowVectorPtr makeCommitOutput(
+    const CommitInfo& commitInfo,
+    const std::string& commitStrategy,
+    velox::memory::MemoryPool* FOLLY_NONNULL pool) {
+  std::vector<VectorPtr> columns = {};
+
+  vector_size_t numOutputRows = 1;
+  FlatVectorPtr<int64_t> rowWrittenVector;
+  FlatVectorPtr<StringView> fragmentsVector;
```

Comment:
fragmentsVector and commitContextVector are used only in the 'else' branch. Consider moving them there.